etcdserver/api/v3rpc/watch.go

```diff
@@ -117,10 +123,30 @@ func (sws *serverWatchStream) recvLoop() error {
 				id = sws.watchStream.Watch(creq.Key, creq.RangeEnd, rev)
 			}
 			sws.ctrlStream <- &pb.WatchResponse{
-				Header:   sws.newResponseHeader(wsrev),
-				WatchId:  int64(id),
-				Created:  true,
-				Canceled: futureRev,
+				Header:        sws.newResponseHeader(wsrev),
+				WatchId:       int64(id),
+				Created:       true,
+				Canceled:      futureRev,
+				StartRevision: rev,
+			}
+
+			if uv.CreateRequest.ProgressReport {
+				go func() {
+					for {
+						select {
+						case <-sws.reportDone:
+							close(sws.reportStream)
+							return
+						case <-time.After(time.Duration(uv.CreateRequest.ReportInterval) * time.Second):
+							sws.reportStream <- &pb.WatchResponse{
```

Comment:
basically return https://github.com/coreos/etcd/blob/master/storage/watchable_store.go#L347 in sync way. this is source of truth.